src/append_session.rs

```diff
@@ -87,83 +87,105 @@ struct AppendState<S>
+    let ack_range = s2_ack.start_seq_num..s2_ack.end_seq_num;
 
-    permit.send(Ok(channel_ack));
+    permit.send(Ok(s2_ack));
 
-    Ok(..end_seq_num)
+    ack_range
 }
 
 async fn resend(
     request_timeout: Duration,
     inflight: &mut VecDeque<InflightBatch>,
     inflight_size: &mut u64,
-    channel_input_tx: mpsc::Sender<types::AppendInput>,
-    channel_ack_stream: &mut ServiceStreamingResponse<AppendSessionStreamingResponse>,
-    last_acked_seqnum: &mut Option<RangeTo<u64>>,
+    s2_input_tx: mpsc::Sender<types::AppendInput>,
+    s2_ack_stream: &mut ServiceStreamingResponse<AppendSessionStreamingResponse>,
+    total_records_acknowledged: &mut usize,
     output_tx: mpsc::Sender<Result<types::AppendOutput, ClientError>>,
 ) -> Result<(), ClientError> {
     debug!(
         inflight_len = inflight.len(),
         inflight_bytes = inflight_size,
         "resending"
     );
+
     let mut resend_index = 0;
     let mut resend_tx_finished = false;
+    let mut stashed_ack = None;
+
     let timer = MuxTimer::<N_TIMER_VARIANTS>::default();
     tokio::pin!(timer);
 
     while !inflight.is_empty() {
         tokio::select! {
-            (event_ord, _) = &mut timer, if timer.is_armed() => {
+            (event_ord, _deadline) = &mut timer, if timer.is_armed() => {
                 match TimerEvent::from(event_ord) {
-                    TimerEvent::BatchDeadline => {
-                        Err(ClientError::Service(Status::cancelled("sdk hit deadline (`request_timeout`) waiting for an append acknowledgement")))?
-                    }
+                    TimerEvent::BatchDeadline => Err(ClientError::Service(Status::cancelled("sdk hit local deadline (`request_timeout`) waiting for an append acknowledgement")))?,
                     _ => unreachable!("only batch deadline timer in resend mode")
                 }
             }
-            Ok(permit) = channel_input_tx.reserve(), if !resend_tx_finished => {
+
+            s2_permit = s2_input_tx.reserve(), if !resend_tx_finished => {
+                let s2_permit = s2_permit.map_err(|_| ClientError::Service(Status::unavailable("server disconnected")))?;
                 match inflight.get(resend_index) {
                     Some(batch) => {
                         timer.as_mut().fire_at(TimerEvent::BatchDeadline, batch.start + request_timeout, CoalesceMode::Earliest);
-                        permit.send(batch.inner.clone());
+                        s2_permit.send(batch.inner.clone());
                         resend_index += 1;
                     },
                     None => resend_tx_finished = true
                 }
             },
-            Some(ack) = channel_ack_stream.next() => {
-                let ack_until = ack_and_pop(
-                    ack?,
+
+            next_ack = s2_ack_stream.next(), if stashed_ack.is_none() => {
+                stashed_ack = Some(next_ack.ok_or(ClientError::Service(Status::internal("s2 stream closed early")))?);
```

Comment:
`s/s2 stream/append response stream` to not conflate

Reply:
I think all such error messages could do with a common prefix that indicates they originate from the client-side, e.g. `client: ...`